## Fake News Twitter Analysis
Consider the 2016 US election Viral Twitter dataset collected between election day (Nov 8th) and March 2017. Tweets have been labelled as containing fake news or not by two sets of people, and where fake news is categorized into one of the five categories: i) Serious fabrication; ii) Large-scale hoaxes; iii) Jokes taken at face value; iv) Slanted reporting of real facts; v) Stories where the 'truth' is contentious. The dataset can be downloaded from Fakenews on 2016 US elections viral tweets (November 2016 - March 2017) | Zenodo

1. Write a script for the calculation of the mean, standard deviation, kurtosis and skewness of Number of follower per user in case of Fake News and Real News dataset. Repeat this process for Favorite Count as well. Conclude whether one can discriminate the two classes using such statistical data.

2. We want to compare the activity of individual users in Fake News and Real News dataset. Select the three most active users in terms of number tweets generated and calculate the average number of tweets generated by the three users. Repeat this process for the five most active users in each dataset, and for the first 10-users in each dataset, and first 15 users for each dataset.

3. We want to compare the average time a user stays before sending a new message. Write a script that uses the date information on the dataset for each tweet to calculate the average waiting time for a random user before sending a new message in case of Fake News and Real News.

## Name: Azizul Kawser
## Student ID: 2207846

In [36]:
# Import libraries
import pandas as pd
import numpy as np
from scipy.stats import kurtosis, skew

In [37]:
# Load dataset
real_news = pd.read_csv("/Users/imak/Documents/Python/Projects/Social Network Analysis/real_news.csv")
real_news['created_at'] = pd.to_datetime(real_news['created_at'])

fake_news = pd.read_csv("/Users/imak/Documents/Python/Projects/Social Network Analysis/fake_news.csv")
fake_news['created_at'] = pd.to_datetime(fake_news['created_at'])

In [38]:
real_news.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2205 entries, 0 to 2204
Data columns (total 19 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   is_fake_news_1             2205 non-null   bool               
 1   is_fake_news_2             2205 non-null   bool               
 2   fake_news_category_1       2205 non-null   int64              
 3   fake_news_category_2       2205 non-null   int64              
 4   tweet_id                   2205 non-null   int64              
 5   created_at                 2205 non-null   datetime64[ns, UTC]
 6   retweet_count              2205 non-null   int64              
 7   text                       2205 non-null   object             
 8   user_screen_name           2205 non-null   object             
 9   user_verified              2205 non-null   bool               
 10  user_friends_count         2205 non-null   int64              
 11  user

In [39]:
fake_news.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 19 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   is_fake_news_1             205 non-null    object             
 1   is_fake_news_2             205 non-null    bool               
 2   fake_news_category_1       205 non-null    int64              
 3   fake_news_category_2       205 non-null    int64              
 4   tweet_id                   205 non-null    int64              
 5   created_at                 205 non-null    datetime64[ns, UTC]
 6   retweet_count              205 non-null    int64              
 7   text                       205 non-null    object             
 8   user_screen_name           205 non-null    object             
 9   user_verified              205 non-null    bool               
 10  user_friends_count         205 non-null    int64              
 11  user_f

In [40]:
fake_news.head()

is_fake_news_1  is_fake_news_2  fake_news_category_1  fake_news_category_2  \
0           True            True                     2                     5   
1           True            True                     5                     5   
2           True            True                     4                     5   
3           True            True                     5                     1   
4           True            True                     1                     5   

             tweet_id                created_at  retweet_count  \
0  587354526857957376 2015-04-12 20:40:06+00:00           1645   
1  620646716849717248 2015-07-13 17:31:23+00:00           1162   
2  644009575025278977 2015-09-16 04:47:02+00:00           1080   
3  654993627098316800 2015-10-16 12:13:44+00:00           7708   
4  655394746630844416 2015-10-17 14:47:38+00:00           3079   

                                                text user_screen_name  \
0  Hillary Clinton has stolen our innovative Wiki...        wikileaks   
1  Services like Uber, Airbnb, and Lyft stimulate...         RandPaul   
2  '@PMgeezer: Only 15 percent of @GOP voters sai...  realDonaldTrump   
3  .@CNN, everyone is now lucidly aware of your i...      colesprouse   
4  .@HillaryClinton is on the front page of the @...  realDonaldTrump   

   user_verified  user_friends_count  user_followers_count  \
0           True                7609               4199734   
1           True                 474               1108061   
2           True                  41              15872757   
3           True                 187               3078353   
4           True                  40              16187377   

   user_favourites_count         tweet_source  geo_coordinates_available  \
0                     31   Twitter Web Client                          0   
1                    741   Twitter Web Client                          0   
2                     45  Twitter for Android                          0   
3                    479   Twitter Web Client                          0   
4                     45   Twitter for iPhone                          0   

   num_hashtags  num_mentions  num_urls  num_media  
0             0             3         0          0  
1             0             2         0          0  
2             0             4         0          0  
3             0             4         0          0  
4             0             3         0          0

<font color='red'> Write a script for the calculation of the mean, standard deviation, kurtosis and skewness of Number of follower per user in case of Fake News and Real News dataset. Repeat this process for Favorite Count as well. Conclude whether one can discriminate the two classes using such statistical data.</font> 

In [41]:
# Calculate the mean, standard deviation, kurtosis, and skewness of the number of followers in the fake news dataset

def get_stat_follower_per_user(df, user):
    """This function will calculate the mean, standard deviation, kurtosis, and skewness of the number of followers
    of a user """
    user_data = df.loc[df['user_screen_name'] == user]
    followers_mean = np.mean(user_data['user_followers_count'])
    followers_std = np.std(user_data['user_followers_count'])
    followers_kurtosis = kurtosis(user_data['user_followers_count'])
    followers_skewness = skew(user_data['user_followers_count'])

    return followers_mean, followers_std, followers_kurtosis, followers_skewness

In [42]:
#For fake news dataset
dataframe = fake_news
user_name = 'wikileaks'
mean, std, kurt, skewness = get_stat_follower_per_user(dataframe, user_name)

print(f'The mean number of follower of {user_name} is', round(mean,2))
print(f'The standard deviation of number of follower of {user_name} is', round(std,2))
print(f'The Kurtosis of number of follower of {user_name} is', round(kurt,2))
print(f'The skewness of number of follower of {user_name} is', round(skewness,2))

The mean number of follower of wikileaks is 4199734.0
The standard deviation of number of follower of wikileaks is 0.0
The Kurtosis of number of follower of wikileaks is -3.0
The skewness of number of follower of wikileaks is 0.0


In [43]:
#For real news dataset
dataframe = real_news
user_name = 'wikileaks'
mean, std, kurt, skewness = get_stat_follower_per_user(dataframe, user_name)

print(f'The mean number of follower of {user_name} is', round(mean,2))
print(f'The standard deviation of number of follower of {user_name} is', round(std,2))
print(f'The Kurtosis of number of follower of {user_name} is', round(kurt,2))
print(f'The skewness of number of follower of {user_name} is', round(skewness,2))

The mean number of follower of wikileaks is 4059615.67
The standard deviation of number of follower of wikileaks is 10584.21
The Kurtosis of number of follower of wikileaks is -1.5
The skewness of number of follower of wikileaks is 0.71


##### Repeating the process for favourite count

In [44]:
def get_stat_favourite_per_user(df, user):
    """This function will calculate the mean, standard deviation, kurtosis, and skewness of the number of followers
    of a user """
    user_data = df.loc[df['user_screen_name'] == user]
    followers_mean = np.mean(user_data['user_favourites_count'])
    followers_std = np.std(user_data['user_favourites_count'])
    followers_kurtosis = kurtosis(user_data['user_favourites_count'])
    followers_skewness = skew(user_data['user_favourites_count'])

    return followers_mean, followers_std, followers_kurtosis, followers_skewness

In [45]:
#For fake news dataset
dataframe = fake_news
user_name = 'wikileaks'
mean, std, kurt, skewness = get_stat_favourite_per_user(dataframe, user_name)

print(f'The mean number of follower of {user_name} is', round(mean,2))
print(f'The standard deviation of number of follower of {user_name} is', round(std,2))
print(f'The Kurtosis of number of follower of {user_name} is', round(kurt,2))
print(f'The skewness of number of follower of {user_name} is', round(skewness,2))

The mean number of follower of wikileaks is 31.0
The standard deviation of number of follower of wikileaks is 0.0
The Kurtosis of number of follower of wikileaks is -3.0
The skewness of number of follower of wikileaks is 0.0


In [46]:
#For real news dataset
dataframe = real_news
user_name = 'wikileaks'
mean, std, kurt, skewness = get_stat_favourite_per_user(dataframe, user_name)

print(f'The mean number of follower of {user_name} is', round(mean,2))
print(f'The standard deviation of number of follower of {user_name} is', round(std,2))
print(f'The Kurtosis of number of follower of {user_name} is', round(kurt,2))
print(f'The skewness of number of follower of {user_name} is', round(skewness,2))

The mean number of follower of wikileaks is 23.67
The standard deviation of number of follower of wikileaks is 0.94
The Kurtosis of number of follower of wikileaks is -1.5
The skewness of number of follower of wikileaks is 0.71


<font color='blue'>From the above statistics of fake news and real news we can conclude that,
- The standard deviation is zero for the fake news class, which means that all followers have the same number.   
- For the real news class, the standard deviation is 10,584.21, which indicates that there is a variation in the number of followers for WikiLeaks.
- The kurtosis is negative for the fake news class and suggests that the distribution of the number of followers is flatter than the normal distribution. The kurtosis of the real news class is also negative, but less so than the fake news class, indicating a slightly flatter distribution.
- The skewness for both classes is close to zero, which suggests that the distributions are roughly symmetrical.

 </font>

<font color='red'> 
We want to compare the activity of individual users in Fake News and Real News dataset. Select the three most active users in terms of number tweets generated and calculate the average number of tweets generated by the three users. Repeat this process for the five most active users in each dataset, and for the first 10-users in each dataset, and first 15 users for each dataset.
 </font>

In [73]:
# Count the number of tweets generated by top n  user

def compare_top_activity_user(df, number_of_user = 3):
    """this function takes the twitter dataframe and number of required most active users.
    and calculate the average number of tweets generated by the users."""
    tweet_counts = df['user_screen_name'].value_counts()
    # Select the three most active users
    top_users = tweet_counts.index[:number_of_user]
    # Filter the dataframe to only include tweets from the top three users
    top_tweets = df[df['user_screen_name'].isin(top_users)]
    # Calculate the average number of tweets generated by the top three users
    print(f"The {number_of_user} most active users are:")
    print(top_users)
    
    for user in top_users:
        user_tweets = top_tweets[top_tweets['user_screen_name'] == user]
        average_tweets = user_tweets.shape[0] / user_tweets['user_screen_name'].nunique()
        print("The average number of tweets generated by", user, "is:", average_tweets)
    
    average_tweets_top_n = top_tweets.groupby('user_screen_name').size().mean()
    print(f"The average number of tweets generated by top {number_of_user} user is : ", round(average_tweets_top_n,1))



In [74]:
# compare the activity of individual users in Fake News of top 3 users.
print('For Fake News dataset: ')
compare_top_activity_user(fake_news)
# compare the activity of individual users in real News of top 3 users.
print('For Real News dataset: ')
compare_top_activity_user(real_news)

For Fake News dataset: 
The 3 most active users are:
Index(['realDonaldTrump', 'JamesOKeefeIII', 'FoxNews'], dtype='object')
The average number of tweets generated by realDonaldTrump is: 22.0
The average number of tweets generated by JamesOKeefeIII is: 5.0
The average number of tweets generated by FoxNews is: 5.0
The average number of tweets generated by top 3 user is :  10.7
For Real News dataset: 
The 3 most active users are:
Index(['realDonaldTrump', 'DanScavino', 'FoxNews'], dtype='object')
The average number of tweets generated by realDonaldTrump is: 292.0
The average number of tweets generated by DanScavino is: 160.0
The average number of tweets generated by FoxNews is: 89.0
The average number of tweets generated by top 3 user is :  180.3


In [75]:
# compare the activity of individual users in Fake News of top 5 users.
print('For Fake News dataset: ')
compare_top_activity_user(fake_news,5)
# compare the activity of individual users in real News of top 5 users.
print('For Real News dataset: ')
compare_top_activity_user(real_news,5)

For Fake News dataset: 
The 5 most active users are:
Index(['realDonaldTrump', 'JamesOKeefeIII', 'FoxNews', 'TeamTrump',
       'NeilTurner_'],
      dtype='object')
The average number of tweets generated by realDonaldTrump is: 22.0
The average number of tweets generated by JamesOKeefeIII is: 5.0
The average number of tweets generated by FoxNews is: 5.0
The average number of tweets generated by TeamTrump is: 4.0
The average number of tweets generated by NeilTurner_ is: 4.0
The average number of tweets generated by top 5 user is :  8.0
For Real News dataset: 
The 5 most active users are:
Index(['realDonaldTrump', 'DanScavino', 'FoxNews', 'mike_pence',
       'elizabethforma'],
      dtype='object')
The average number of tweets generated by realDonaldTrump is: 292.0
The average number of tweets generated by DanScavino is: 160.0
The average number of tweets generated by FoxNews is: 89.0
The average number of tweets generated by mike_pence is: 81.0
The average number of tweets generated by

In [81]:
# Count the number of tweets generated by first n  user

def compare_first_n_activity_user(df, number_of_user = 10):
    """this function takes the twitter dataframe and number of required most active users.
    and calculate the average number of tweets generated by the users."""
    # tweet_counts = df['user_screen_name'].value_counts()
    # Select the three most active users
    top_users = df['user_screen_name'].head(number_of_user)
    # Filter the dataframe to only include tweets from the top three users
    top_tweets = df[df['user_screen_name'].isin(top_users)]
    # Calculate the average number of tweets generated by the top three users
    print(f"The firtst {number_of_user}  users are:")
    print(top_users)
    
    for user in top_users:
        user_tweets = top_tweets[top_tweets['user_screen_name'] == user]
        average_tweets = user_tweets.shape[0] / user_tweets['user_screen_name'].nunique()
        print("The average number of tweets generated by", user, "is:", average_tweets)
    
    average_tweets_top_n = top_tweets.groupby('user_screen_name').size().mean()
    print(f"The average number of tweets generated by first {number_of_user} user is : ", round(average_tweets_top_n,1))

In [82]:
# for the first 10-users in each dataset
print('For Fake News dataset: ')
compare_first_n_activity_user(fake_news)
print('For real News dataset: ')
compare_first_n_activity_user(real_news)

For Fake News dataset: 
The firtst 10  users are:
0          wikileaks
1           RandPaul
2    realDonaldTrump
3        colesprouse
4    realDonaldTrump
5         DaveWeasel
6         SteveGuest
7         DanScavino
8          BBCTaster
9    realDonaldTrump
Name: user_screen_name, dtype: object
The average number of tweets generated by wikileaks is: 1.0
The average number of tweets generated by RandPaul is: 2.0
The average number of tweets generated by realDonaldTrump is: 22.0
The average number of tweets generated by colesprouse is: 1.0
The average number of tweets generated by realDonaldTrump is: 22.0
The average number of tweets generated by DaveWeasel is: 1.0
The average number of tweets generated by SteveGuest is: 1.0
The average number of tweets generated by DanScavino is: 4.0
The average number of tweets generated by BBCTaster is: 1.0
The average number of tweets generated by realDonaldTrump is: 22.0
The average number of tweets generated by first 10 user is :  4.1
For real Ne

In [83]:
# and first 15 users for each dataset.
print('For Fake News dataset: ')
compare_first_n_activity_user(fake_news,15)
print('For real News dataset: ')
compare_first_n_activity_user(real_news,15)

For Fake News dataset: 
The firtst 15  users are:
0           wikileaks
1            RandPaul
2     realDonaldTrump
3         colesprouse
4     realDonaldTrump
5          DaveWeasel
6          SteveGuest
7          DanScavino
8           BBCTaster
9     realDonaldTrump
10    realDonaldTrump
11         katya_zamo
12    HomunculusLoikm
13            JebBush
14      EntheosShines
Name: user_screen_name, dtype: object
The average number of tweets generated by wikileaks is: 1.0
The average number of tweets generated by RandPaul is: 2.0
The average number of tweets generated by realDonaldTrump is: 22.0
The average number of tweets generated by colesprouse is: 1.0
The average number of tweets generated by realDonaldTrump is: 22.0
The average number of tweets generated by DaveWeasel is: 1.0
The average number of tweets generated by SteveGuest is: 1.0
The average number of tweets generated by DanScavino is: 4.0
The average number of tweets generated by BBCTaster is: 1.0
The average number of tw